source/fab/builder.py

```diff
@@ -119,43 +109,67 @@ def __init__(self,
-            )
-        if ld_flags != '':
-            self._command_flags_map[FortranLinker] = (
-                ld_flags.split()
-            )
-        self._queue = QueueManager(n_procs - 1)
 
-    def _extend_task_queue(self, task: Task) -> None:
-        self._queue.add_to_queue(task)
-        for prereq in task.prerequisites:
-            self._queue.add_to_queue(HashCalculator(FileTextReader(prereq),
-                                                    self._state))
+        # If not given, use the name of the target as the exec name
+        # TODO: Should this be in the CLI part?
+        if exec_name == '':
+            exec_name = target
+
+        # Path maps tell the engine what filetype and starting state
+        # the Artifacts representing any files encountered by the
+        # initial descent should have
+        path_maps = [
+            PathMap(r'.*\.f90', FortranSource, Raw),
+            PathMap(r'.*\.F90', FortranSource, Seen),
+        ]
+
+        # Initialise the required Tasks, providing them with any static
+        # properties such as flags to use, workspace location etc
+        # TODO: Eventually the tasks may instead access many of these
+        # properties via the configuration (at Task runtime, to allow for
+        # file-specific overrides?)
+        fortran_preprocessor = FortranPreProcessor(
+            'cpp', ['-traditional-cpp', '-P'] + fpp_flags.split(), workspace
+        )
+        fortran_analyser = FortranAnalyser(workspace)
+        fortran_compiler = FortranCompiler(
+            'gfortran',
+            ['-c', '-J', str(workspace)] + fc_flags.split(), workspace
+        )
+        # TODO: Linker class is probably generic not Fortran specific
+        # now that we are passing in the command to use?
+        fortran_linker = FortranLinker(
+            'gfortran', ld_flags.split(), workspace, exec_name
+        )
+
+        # The Task map tells the engine what Task it should be using
+        # to deal with Artifacts depending on their type and state
+        task_map = {
+            (FortranSource, Seen): fortran_preprocessor,
+            (FortranSource, Raw): fortran_analyser,
+            (FortranSource, Analysed): fortran_compiler,
+            (BinaryObject, Compiled): fortran_linker,
+        }
+
+        engine = Engine(workspace,
+                        target,
+                        path_maps,
+                        task_map)
+        self._queue = QueueManager(n_procs - 1, engine)
+
+    def _extend_queue(self, artifact: Artifact) -> None:
+        self._queue.add_to_queue(artifact)
 
     def run(self, source: Path):
 
         self._queue.run()
 
-        visitor = SourceVisitor(self._precompile_map,
-                                self._command_flags_map,
-                                self._state,
-                                self._workspace,
-                                self._extend_task_queue)
+        visitor = SourceVisitor(self._extend_queue)
```

Comment:
Yes, reducing number of arguments to the tree descender feels right.

Reply:
Do you think it's still worth having the callback (now that the callback only passes its arguments straight on)?  I deliberated over passing the queue instead and calling it directly but decided to keep it on the offchance we want to add some sort of additional processing to it later

Reply:
I think we retain the callback for the moment.